federation: Updating KubeDNS to try finding a local service first for federation query

nikhiljindal opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/26762

Updating KubeDNS to try to find a local service first for federation query.
Without this change, KubeDNS always returns the DNS hostname, even if a local service exists.

Have updated the code to first remove federation name from path if it exists, so that the default search for local service happens. If we dont find a local service, then we try to find the DNS hostname.

Will appreciate a strong review since this is my first change to KubeDNS.
https://github.com/kubernetes/kubernetes/pull/25727 was the original PR that added federation support to KubeDNS.

cc @kubernetes/sig-cluster-federation @quinton-hoole @madhusudancs @bprashanth @mml

nikhiljindal committed:
Updating KubeDNS to try a local service first for federation query

nikhiljindal commented on pkg/dns/dns.go:
```diff
@@ -420,8 +420,20 @@ func (kd *KubeDNS) newHeadlessService(service *kapi.Service) error {
+	path := strings.Split(trimmed, ".")
+	isFederationQuery := false
+	federationPath := []string{}
+	if !exact && kd.isFederationQuery(path) {
```
I am not sure about this exact parameter.
Looking at the existing code, it seems that we should be checking !exact here.
Can remove it if its not required.

quinton-hoole replied:
@nikhiljindal See https://github.com/kubernetes/kubernetes/pull/25727#discussion_r63759326 for context.

quinton-hoole commented:
Thanks @nikhiljindal - I'm going to review this now.

madhusudancs commented on pkg/dns/dns.go:
```diff
@@ -420,8 +420,20 @@ func (kd *KubeDNS) newHeadlessService(service *kapi.Service) error {
 	trimmed := strings.TrimRight(name, ".")
-	segments := strings.Split(trimmed, ".")
-	path := reverseArray(segments)
+	path := strings.Split(trimmed, ".")
```
Call this segments for terminology consistency. path is the reverse segments.

madhusudancs commented:
LGTM other than a minor nit. Feel free to apply the label after fixing it.

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -420,8 +420,20 @@ func (kd *KubeDNS) newHeadlessService(service *kapi.Service) error {
+		federationPath = append(federationPath, path...)
+		// To try local service, remove federation name from path.
+		// Federation name is 3rd in the path (after service name and namespace).
+		path = append(path[:2], path[3:]...)
```
@nikhiljindal @madhusudancs I'm probably overlooking something, but its not clear to me where we ensure that we avoid returning the cluster IP of a local service that has no healthy endpoints backing it?

nikhiljindal replied:
It works when there is no endpoint for the local service (I have a test for it).
But I just tried deleting endpoints from an existing service and it still returned me the endpoint (which I deleted). Am debugging that.

nikhiljindal replied:
Well that was just an error in the way I was testing.
What I said above was only true for headless services. Updated the code to handle normal services as well (which is what federation will create)

nikhiljindal commented:
Updated the code to check for endpoints of a service before returning its cluster IP for a federation query.
We return the DNS hostname instead if there are no endpoints.

Added tests as well.
I have also added a test for headless service (since I wrote that first), but realized that we wont be creating headless services in federation. So its not strictly required, but I have let it be.

PTAL

quinton-hoole commented:
This is a pretty substantial change that will need to be carefully reviewed and tested.  If we break KubeDNS, the cluster is broken, so that's really bad.  Do we understand why it appeared to work, at least in a limited fashion, before these changes?

To be clear, I've not reviewed to code extensively yet.  But I'm very nervous about making changes to such a key component of every Kubernetes cluster so late before a major release, unless we are super-confident about the proposed changes.

Are we super-confident @nikhiljindal @madhusudancs ?  @ArtfulCoder could you also please take a look?  More eyes are better.

quinton-hoole commented:
PS: I would ideally like us to bang hard on this for a few days in a soak test cluster before declaring it good.  I can assist in making that happen.

nikhiljindal commented:
Totally agree @quinton-hoole 
This is the first time I am making a change to kubeDNS so I am definitely not super confident :)

I have restricted the changes to be only when the query is for federation service. And have added extensive tests. The existing code has good tests as well.
But yes, this is a critical component.

Will wait until someone familiar with this code has had a chance to take a look.
@bprashanth maybe?

nikhiljindal commented:
@k8s-bot node e2e test this issue: https://github.com/kubernetes/kubernetes/issues/27668

nikhiljindal commented:
cc @girishkalele who I am told is the new KubeDNS owner :)

nikhiljindal commented:
Added a check to validate that federation names dont have dots '.' as per discussion with @ArtfulCoder and @madhusudancs

ArtfulCoder commented:
TODOs left in the PR:
1. make sure its thread safe when the new map is accessed.
2. removed the reverse lookup map
3. ensure that federation name does not have dots.
4. check for wildcard support and impact on wildcards with the new federation segment.
Consider scenarios where the federation name could be an existing namespace/service name.

quinton-hoole commented:
@nikhiljindal Still needs rebase, and the verification error is below.  Hopefully the rebase will sort that out too.

```
FAIL: changes needed but not made due to --verify
/go/src/k8s.io/kubernetes/docs/ is out of date. Please run hack/update-munge-docs.sh
FAILED   ./hack/../hack/verify-munge-docs.sh	2s
```

nikhiljindal committed:
Updating the Makefile and YAML files to use the updated image

nikhiljindal commented:
Rebased and updated the code as per comments.

> TODOs left in the PR:
> 1. make sure its thread safe when the new map is accessed.

Added a lock as discussed.

> 2. removed the reverse lookup map

Have added a TODO to remove it.

> 3. ensure that federation name does not have dots.

Added validation

> 4. check for wildcard support and impact on wildcards with the new federation segment.

Added a comment stating that we dont support wildcard federation queries. We can add support for that later, if required.

> Consider scenarios where the federation name could be an existing namespace/service name.

Yea. Need to think a bit more about the case when federation name is same as namespace name. Will get back on this.

quinton-hoole commented:
@nikhiljindal Node e2e test failure looks like a flake. 
@k8s-bot test this issue: #IGNORE 

```
â€¢ Failure [5.027 seconds]
[k8s.io] Container Runtime Conformance Test
/var/lib/jenkins/workspace/node-pull-build-e2e-test/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:685
  container runtime conformance blackbox test
  /var/lib/jenkins/workspace/node-pull-build-e2e-test/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:279
    when running a container with a new image
    /var/lib/jenkins/workspace/node-pull-build-e2e-test/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:278
      should not be able to pull non-existing image from gcr.io [It]
      /var/lib/jenkins/workspace/node-pull-build-e2e-test/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e_node/runtime_conformance_test.go:276

      Expected error:
          <*errors.errorString | 0xc820965c80>: {
              s: "unexpected container statuses []",
          }
          unexpected container statuses []
      not to have occurred

```

quinton-hoole commented:
Flake tracking bug is: https://github.com/kubernetes/kubernetes/issues/23971

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -458,13 +509,41 @@ func (kd *KubeDNS) Records(name string, exact bool) (retval []skymsg.Service, er
 }
 
+// Returns true if the service corresponding to the given message has endpoints.
+// Always returns true for headless services.
```
I don't understand why you want to return true for headless services, even if they have no endpoints?  Please explain.

nikhiljindal replied:
updated code

quinton-hoole commented on pkg/dns/dns.go:
```diff
@@ -558,6 +637,9 @@ func getSkyMsg(ip string, port int) (*skymsg.Service, string) {
 //   7. And federation must be one of the listed federations in the config.
+//   Note: Because of the above conditions, this method will treat wildcard queries such as
+//   *.mysvc.myns.myfederation.svc.domain.path as non-federation queries.
+//   We can add support for wildcard queries later, if needed.
```
Yes, I would like us to support wildcard queries for federated services, as this will be an easy way for clients to discover all shared of a service globally, with a single DNS query.  Pretty cool.  Can wait until post-1.3.

madhusudancs replied:
How are the clients expected to format the wildcard-names? Specifically, where do you expect them to put the `*`?

madhusudancs replied:
Also, what do you expect the KubeDNS response to be?

nikhiljindal commented:
Tests on this PR ran with the updated KubeDNS image (with my code changes). Great to see that they all passed.

> > Consider scenarios where the federation name could be an existing namespace/service name.
> 
> Yea. Need to think a bit more about the case when federation name is same as namespace name. Will get back on this.

I discussed this with @quinton-hoole and we discussed a few potential solutions, but decided not to fix it in 1.3. Have filed https://github.com/kubernetes/kubernetes/issues/27969 to keep track.


Also, @quinton-hoole suggested returning a CNAME record for the local service (if it exists) instead of the A records directly in response to a federation DNS query. Making that change now.

quinton-hoole commented on pkg/dns/dns_test.go:
```diff
@@ -350,6 +360,98 @@ func TestHeadlessServiceWithDelayedEndpointsAddition(t *testing.T) {
+	assert.Equal(t, a, records[0].Host)
+}
+
+// Verifies that quering KubeDNS for a headless federation service returns the DNS hostname when a local service does not exist and returns the endpoint IP when a local service exists.
```
As discussed, for both headless and non-headless, we should return a CNAME (to the local service's DNS name if the service exists and has endoints, or to the zone-local DNS name if the local service has no endpoints).

nikhiljindal replied:
done

yifan-gu commented:
@nikhiljindal Could you update the manifest under `cluster/gce/coreos/kube-manifest/addons` as well? Thank you!

nikhiljindal commented:
Sure @yifan-gu done. 

Update code as per comments. Added a commit to return the DNS hostname of local service rather than returning its endpoints directly. PTAL.

nikhiljindal commented:
Have also rebased to include other DNS changes, specifically: https://github.com/kubernetes/kubernetes/pull/27891, https://github.com/kubernetes/kubernetes/pull/27896 and https://github.com/kubernetes/kubernetes/pull/27960.

Pushed the updated image to gcr.io/google_containers/kubedns-amd64:1.5 to run tests

girishkalele commented on pkg/dns/dns.go:
```diff
@@ -422,7 +439,76 @@ func (kd *KubeDNS) Records(name string, exact bool) (retval []skymsg.Service, er
+	isFederationQuery := false
+	federationSegments := []string{}
+	if !exact && kd.isFederationQuery(segments) {
+		glog.Infof("federation service query: Received federation query. Going to try to find local service first")
```
Move this to a lower log level  ? Seems chatty with no specific information.

nikhiljindal replied:
the information this gives is that this query was interpreted as a federation query.
I would like to keep it for debugging.
Note that we wont see this in the logs unless someone has federation enabled and is sending federation queries.

girishkalele commented on pkg/dns/dns.go:
```diff
@@ -422,7 +439,76 @@ func (kd *KubeDNS) Records(name string, exact bool) (retval []skymsg.Service, er
+	validRecord := false
+	for _, val := range records {
+		if kd.isHeadlessServiceRecord(&val) {
+			// We know that a headless service has endpoints for sure if a record was returned for it.
```
Nit: Can invert the if check and eliminate this empty if block.

nikhiljindal replied:
fixed

nikhiljindal commented:
I see a new GKE e2e here which is failing with some auth problems unrelated to this PR. I dont see any issue filed. 
cc @fejta ?

girishkalele commented:
@nikhiljindal 

Both gke e2e seem to be known flakes.
Is this the kubernetes-e2e-gke-slow/5723 run ?

```
[Fail] [k8s.io] Pod Disks [It] should schedule a pod w/ a readonly PD on two hosts, then remove both. [Slow] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/pd.go:172
```

or kubernetes-e2e-gke/9788 ?

```
[Fail] [k8s.io] Proxy version v1 [It] should proxy through a service and a pod [Conformance] 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/proxy.go:259
```

girishkalele commented:
As far as I can tell, there is no non-federation functionality affected.
2 Major comments:
1. Logs are very noisy, I ran against kubedns:1.5 and see very chatty logs. Please reduce log levels.
2. Squash all commits.

nikhiljindal commented:
Squashed into 2 commits (dns changes and YAML file changes).
As discussed the logs are same as in 1.4

girishkalele commented:
LGTM - lets get it in.

nikhiljindal commented:
thx @girishkalele

quinton-hoole commented:
@nikhiljindal I suggest that you follow @girishkalele 's suggestions re logging verbosity.  Rather log as glog.V(4) and set --v=4 on the binary while you're debugging?  Will that work for you?

quinton-hoole commented:
@nikhiljindal @fejta PS that new GKE smoke test is failing on #27972 also.

nikhiljindal commented:
Pushed and verified that 1.5 image now exists on all platforms:
https://gcr.io/v2/google_containers/kubedns-arm64/tags/list
https://gcr.io/v2/google_containers/kubedns-ppc64le/tags/list
https://gcr.io/v2/google_containers/kubedns-arm/tags/list
https://gcr.io/v2/google_containers/kubedns-amd64/tags/list
